src/subscription.ts

```diff
@@ -1,263 +1,71 @@
-                }
-                let data = JSON.parse(rawEvent);
-                let err = this.onMessage(data);
-                if (err != null) {
-                    return err;
-                }
-            }
-        }
-    }
-
-    // calls options.onEvent 0+ times, then returns an Error or null
-    private onMessage(message: any[]): Error {
-        this.assertState(['OPEN']);
-
-        if (this.gotEOS) {
-            return new Error("Got another message after EOS message");
-        }
-        if (!Array.isArray(message)) {
-            return new Error("Message is not an array");
-        }
-        if (message.length < 1) {
-            return new Error("Message is empty array");
-        }
-
-        switch (message[0]) {
-            case 0:
-                return null;
-            case 1:
-                return this.onEventMessage(message);
-            case 255:
-                return this.onEOSMessage(message);
-            default:
-                return new Error("Unknown Message: " + JSON.stringify(message));
-        }
-    }
-
-    // EITHER calls options.onEvent, OR returns an error
-    private onEventMessage(eventMessage: any[]): Error {
-        this.assertState(['OPEN']);
-
-        if (eventMessage.length !== 4) {
-            return new Error("Event message has " + eventMessage.length + " elements (expected 4)");
-        }
-        let [_, id, headers, body] = eventMessage;
-        if (typeof id !== "string") {
-            return new Error("Invalid event ID in message: " + JSON.stringify(eventMessage));
-        }
-        if (typeof headers !== "object" || Array.isArray(headers)) {
-            return new Error("Invalid event headers in message: " + JSON.stringify(eventMessage));
-        }
-
-        if (this.options.onEvent) { this.options.onEvent({ eventId: id, headers: headers, body: body }); }
-    }
-
-    // calls options.onEvent 0+ times, then possibly returns an error
-    private onEOSMessage(eosMessage: any[]): Error {
-        this.assertState(['OPEN']);
+export interface SubscriptionState {
+    unsubscribe();
+}
 
-        if (eosMessage.length !== 4) {
-            return new Error("EOS message has " + eosMessage.length + " elements (expected 4)");
-        }
-        let [_, statusCode, headers, info] = eosMessage;
-        if (typeof statusCode !== "number") {
-            return new Error("Invalid EOS Status Code");
-        }
-        if (typeof headers !== "object" || Array.isArray(headers)) {
-            return new Error("Invalid EOS Headers");
-        }
+export interface SubscriptionEvent {
+    eventId: string;
+    headers: ElementsHeaders;
+    body: any;
+}
 
-        this.state = SubscriptionState.ENDING;
-        return new ErrorResponse(statusCode, headers, info);
-    }
+export interface SubscriptionListeners {
+    onOpen?: (headers: ElementsHeaders) => void;
+    onSubscribe?: () => void;
+    onRetrying?:() => void;
+    onResuming?:() => void;
```

Comment:
I see a slight problem here - the `SubscriptionListeners` should represent listeners for generic subscriptions, but some listeners for concrete implementations sneaked in, like `onRetrying` and `onResuming`. I think they should be moved to separate interfaces - `RetryingSubscriptionListeners` and `ResumingSubscriptionListeners`.

Reply:
I think I'll remove `onResuming` and use `onRetrying` for both cases. I think it makes the implementation easier that way, and it can still change if need be.